### PR TITLE
Update to the new Observability extension model

### PR DESCRIPTION
**Pull request by nadundesilva**
Related to #26217

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java [hunk 1]
```diff
@@ -122,10 +125,31 @@ private static void loadConfigurations(Map<String, String> configArgs, String co
                 ((BLogManager) logManager).loadUserProvidedLogConfiguration();
             }
 
-            boolean observeFlag = ConfigRegistry.getInstance().getAsBoolean(CONFIG_OBSERVABILITY_ENABLED);
-            if (observeFlag) {
-                ConfigRegistry.getInstance().addConfiguration(CONFIG_METRICS_ENABLED, Boolean.TRUE);
-                ConfigRegistry.getInstance().addConfiguration(CONFIG_TRACING_ENABLED, Boolean.TRUE);
+            if (ConfigRegistry.getInstance().contains(CONFIG_OBSERVABILITY_ENABLED)) {
```
Comment (rdhananjaya) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java:128]: Shall we just extract a variable for this `ConfigRegistry.getInstance()`

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java [hunk 2]
```diff
@@ -122,10 +125,31 @@ private static void loadConfigurations(Map<String, String> configArgs, String co
...
+                // Only try to inherit if "b7a.observability" config is provided
+                boolean observeFlag = ConfigRegistry.getInstance().getAsBoolean(CONFIG_OBSERVABILITY_ENABLED);
```
Comment (rdhananjaya) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java:130]: I think this is always `true`

**Review by rdhananjaya — APPROVED** [review]
Compiler FE changes looks good.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java (hunk 2 above)
Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java:130]: Can the key exist and value be `false` in config registry ?
A user can set it and that is why I changed it to this way from previous implementation.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java (hunk 1 above)
Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java]: Fixed in 0420a885f7cd6f85221a5204378c0c3f791cf5ec

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java (hunk 2 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/launch/LaunchUtils.java]: Ahh got it, thanks.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 3]
```diff
@@ -167,6 +168,15 @@ boolean isDefaultModule() {
             moduleLoadRequests.addAll(docContext.moduleLoadRequests(PackageDependencyScope.DEFAULT));
         }
 
+        // TODO: Move to compiler extension once new Compiler Extension model is introduced
```
Comment (hevayo) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:171]: Should this be added to every module ? I guess it only need to be added to one ?

Comment (nadundesilva) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:171]: Yeah when building a package, if `observability-included` is set to true, we need to add it to one of the modules in the package.

Shall I move it to PackageResolution class and add it once ?

Comment (nadundesilva) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java]: Fixed in f156a5995066afd87a6fe03c1cdce62c78ecd0bf

**Comment by nadundesilva** [comment]
Merged ballerina-lang/master to validate whether all tests are passing with latest changes.

**Review by hevayo — APPROVED** [review]

File: observelib/observe-internal/src/main/java/org/ballerinalang/observe/NativeFunctions.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    public static boolean isMetricsEnabled() {
+        return ObserveUtils.isMetricsEnabled();
+    }
+
+    public static boolean isTracingEnabled() {
+        return ObserveUtils.isTracingEnabled();
+    }
+
+    public static BError enableMetrics() {
+        ConfigRegistry configRegistry = ConfigRegistry.getInstance();
```
Comment (manuranga) [inline observelib/observe-internal/src/main/java/org/ballerinalang/observe/NativeFunctions.java:57]: config api will be deprecated in the future, will the be a way to set this then?

Comment (nadundesilva) [reply observelib/observe-internal/src/main/java/org/ballerinalang/observe/NativeFunctions.java:57]: Will there be a java level API ? If not I guess we will have to pass it in the Java interop call for now.

Comment (nadundesilva) [reply observelib/observe-internal/src/main/java/org/ballerinalang/observe/NativeFunctions.java:57]: Created an issue #27846 to refactor the code for configuration reading to support the new configuration feature.

**Pull request merged**